delfin/drivers/dell_emc/vmax/perf_utils.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2020 The SODA Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+
+# minimum interval supported by VMAX
+VMAX_PERF_MIN_INTERVAL = 5
+
+ARRAY_METRICS = ["HostIOs",
+                 "HostMBWritten",
+                 "ReadResponseTime",
+                 "HostMBReads",
+                 "HostReads",
+                 "HostWrites",
+                 "WriteResponseTime"
+                 ]
+
+VMAX_REST_TARGET_URI_ARRAY_PERF = '/performance/Array/metrics'
+
+
+def epoch_time_ms_now():
+    """Get current time in epoch ms.
+    :returns: epoch time in milli seconds
+     """
+    ms = int(time.time() * 1000)
+    return ms
+
+
+def epoch_time_interval_ago(interval_seconds=VMAX_PERF_MIN_INTERVAL):
+    """Get epoch time in milliseconds  before an interval
+    :param interval_seconds: interval in seconds
+    :returns: epoch time in milliseconds
+    """
+    return int(epoch_time_ms_now() - (interval_seconds * 1000))
+
+
+def generate_performance_payload(array, interval, metrics):
+    """Generate request payload for VMAX performance POST request
+    :param array: symmetrixID
+    :param interval: interval in seconds
+    :returns: payload dictionary
+    """
+    return {'symmetrixId': str(array),
+            "endDate": epoch_time_ms_now(),
+            "startDate": epoch_time_interval_ago(interval),
+            "metrics": metrics,
+            "dataFormat": "Average"}
+
+
+def parse_performance_data(response):
+    """Parse metrics response to a map
+    :param response: response from unispshere REST API
+    :returns: map with key as metric name and value as dictionary
+        containing {timestamp: value} for a the timestamps available
+    """
+    metrics_map = {}
+    for metrics in response["resultList"]["result"]:
```

Comment:
Is it not possible to combine parse_performance_data() and map_array_perf_metrics_to_delfin_metrics() into same. We already have a list of keys in ARRAY_METRICS
Here we are creating a map just to break it in the second func call

Reply:
parse_performance_data() is generic , can be used for any resource performnce data parsing.
map_array_perf_metrics_to_delfin_metrics()  is only to map array leval metrics of VMAX to delfin array leval metic model. This is why kept it as separate.

Reply:
ok